src/main/java/org/kuali/kfs/coa/document/validation/impl/AccountGlobalRule.java

```diff
@@ -863,5 +963,959 @@ private boolean isIncomeStreamAccountRequired(String fundGroupCode, String subFu
+					activeICRList.add(icrAccount);
+				}
+			}
+			success = !activeICRList.isEmpty();
+			if (!success) {
+				putFieldError(KFSPropertyConstants.INDIRECT_COST_RECOVERY_ACCOUNTS, CUKFSKeyConstants.ERROR_DOCUMENT_ACCT_GLB_MAINT_ICR_EMPTY_FOR_CG_ACCOUNT, new String[] { subFundGroupCode, accountGlobalDetail.getAccountNumber() });
+			}
+		}
+
+		return success;
+	}
+    
+    protected boolean checkICRCollectionDoesNotExistForNonCGSubFund(CuAccountGlobal newAccount, String subFundGroupCode){
+    	boolean success = true;
+    	success = newAccount.getActiveIndirectCostRecoveryAccounts().isEmpty();
+
+    	// if any active ICR accts entered then put error message; no need to check acct details
+    	if (!success){
+    		putFieldError(KFSPropertyConstants.INDIRECT_COST_RECOVERY_ACCOUNTS, KFSKeyConstants.ERROR_DOCUMENT_ACCMAINT_CG_ICR_FIELDS_FILLED_FOR_NON_CG_ACCOUNT,newAccount.getSubFundGroupCode());
+    	}
+
+    	// if no active ICR accounts on Global doc then check acct details
+    	if(success){
+    		if(ObjectUtils.isNotNull(newAccount.getAccountGlobalDetails()) && !newAccount.getAccountGlobalDetails().isEmpty()){
+    			for(AccountGlobalDetail accountGlobalDetail : newAccount.getAccountGlobalDetails()){
+    				accountGlobalDetail.refreshReferenceObject(KFSPropertyConstants.ACCOUNT);
+    				List<IndirectCostRecoveryAccount> activeICRList = new ArrayList<IndirectCostRecoveryAccount>();
+
+    				for(IndirectCostRecoveryAccount icrAccount : accountGlobalDetail.getAccount().getActiveIndirectCostRecoveryAccounts()){
+    					boolean foundInactivated = false;
+    					for(IndirectCostRecoveryAccountChange icrAccountChange : newAccount.getIndirectCostRecoveryAccounts()){
+    						if(!icrAccountChange.isActive()){
+    							if(icrAccount.getIndirectCostRecoveryFinCoaCode().equalsIgnoreCase(icrAccountChange.getIndirectCostRecoveryFinCoaCode()) && icrAccount.getIndirectCostRecoveryAccountNumber().equalsIgnoreCase(icrAccountChange.getIndirectCostRecoveryAccountNumber())){
+    								foundInactivated = true;
+    								break;
+    							}
+    						}
+    					}
+    					if(!foundInactivated){
+    						activeICRList.add(icrAccount);
+    					}
+    				}
+    				success &= activeICRList.isEmpty();
+    				if (!success){
+    					putFieldError(KFSPropertyConstants.INDIRECT_COST_RECOVERY_ACCOUNTS, CUKFSKeyConstants.ERROR_DOCUMENT_ACCT_GLB_MAINT_ICR_NOT_EMPTY_FOR_NON_CG_ACCOUNT, new String[]{subFundGroupCode, accountGlobalDetail.getAccountNumber()});
+    				}
+    			}
+    		}
+    	}
+    	
+    	return success;	
+    }
+    
+    protected boolean checkICRCollectionDoesNotExistForNonCGSubFund(CuAccountGlobal newAccount, AccountGlobalDetail accountGlobalDetail, String subFundGroupCode){
+    	boolean success = true;
+    	success = newAccount.getActiveIndirectCostRecoveryAccounts().isEmpty();
+
+    	// if any active ICR accts entered then put error message; no need to check acct details
+    	if (!success){
```

Comment:
Thank you for pointing this out. I will refactor this code because it is not very clear and I think the current implementation for matching new with existing ICR accounts is incorrect as it does not take into account the percentage. I will update this and retest this functionality.